### PR TITLE
Allow authors to define permissions for each node

### LINKED ISSUE #32: Allow authors to define permissions for each node
In the UI, allow the author to define the permission for each node.

Note: The 'read' permission will affect the node itself AND recursively any child nodes that need this node to be attached to the tapestry. In other words, user cannot see the children of this node or any of its grand-children, grand-grand-children, etc. if the read flag is not on for this node.

* The permission is defined for each group:
  * Groups are
    * Public (default, always there)
    * Group (used to give permission to a group)
    * User (used to give permission to a user)
  * Options for permissions (select 0 or more):
    * Read (default)
    * Edit-Submit (allows user(s) to edit their own version of the given node, and submit it to update the main tapestry)
    * Add-Submit (allows user(s) to attach new nodes to that node in their own version of the tapestry, and submit it to be added to the main tapestry)
    * Delete (allows user(s) to delete this node in their own version of the tapestry, and submit it to be deleted from the main tapestry)

### PR DESCRIPTION
**Pull request by anhduc130**
Fixes #32

### CODE REVIEW

File: controller/class.tapestry-controller.php [hunk 1]
```diff
@@ -337,6 +401,7 @@ private function _makeMetadata($node, $nodePostId)
         return (object)array(
             'post_id' => $nodePostId,
```
[the remark this answers is not in the document]
Comment (wynnset) [reply controller/class.tapestry-controller.php]: Please also do the same with you doc comment parameters. Align the parameter names together and the parameter descriptions together.

File: controller/class.tapestry-controller.php [hunk 2]
```diff
@@ -96,6 +134,39 @@ public function addTapestryNode($node)
...
+        }
+
+        if (!$this->_isChildNodeOfTapestry($nodeMetaId)) {
+            return $this->_throwsError('INVALID_CHILD_NODE');
+        }
+
+        // TODO: validate that $permissions has appropriate/valid info
+
+        $nodeMetadata = get_metadata_by_mid('post', $nodeMetaId)->meta_value;
+
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Please remove these extra lines between all your code so they look more grouped together.

File: controller/class.tapestry-controller.php [hunk 3]
```diff
@@ -10,18 +10,37 @@ class TapestryController
         'TAPESTRY' => 'tapestry',
         'TAPESTRY_NODE' => 'tapestry_node'
     ];
+    const NODE_PERMISSIONS = [
+        'DEFAULT' => [
+            'public' => 'read'
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: I think this needs to be an array like we discussed yesterday.

File: controller/class.tapestry-controller.php [hunk 4]
```diff
@@ -10,18 +10,37 @@ class TapestryController
...
+        ],
+        'OPTIONS' => [
+            'READ_NODE' => 'read',
+            'EDIT_SUBMIT_NODE' => 'edit_submit',
+            'ADD_SUBMIT_NODE' => 'add_submit',
+            'DELETE_NODE' => 'delete'
```
Comment (wynnset) [inline controller/class.tapestry-controller.php]: Please add `'edit'`, `'add'`, and `'approve'` permissions. `'delete'` permission should be removed. Anyone who has `'edit'` permission should be also able to delete. `'edit_submit'` should be able to submit deletion that goes through a similar approval process.

File: controller/class.tapestry-controller.php [hunk 5]
```diff
@@ -96,6 +130,36 @@ public function addTapestryNode($node)
         return $node;
     }
 
+    /**
+     * Update Tapestry Node Permissions
+     * 
+     * @param Integer $nodeMetaId Node meta id
```
[the remark this answers is not in the document]
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: Replaced with tabs.

File: controller/class.tapestry-controller.php [hunk 6]
```diff
@@ -155,7 +217,7 @@ public function updateTapestrySettings($settings)
     /**
      * Retrieve a Tapestry post
      * 
-     * @return Object $tapestry
+     * @return  Object  $tapestry
```
[the remark this answers is not in the document]
Comment (anhduc130) [reply controller/class.tapestry-controller.php]: @wynnset I will resolve those kinds of comments because we agree to align the params with tabs.

**Pull request merged**